### ISSUE TITLE
Some clang-tidy warnings get struck-through

### ISSUE THREAD

**Issue by arijun** [clangd/vscode-clangd#482]
Some clang-tidy warnings are displayed with a strikethrough. From what I can tell, this happens with modernize- lints but not e.g. bugprone- or unused-includes. This is a recent development, previously they just had an orange underline. This, combined with the expansive way the error can be defined, can make it hard to read some things (crossing out all hundred lines of a `typedef`ed enum makes it pretty hard to skim).

![image](https://github.com/clangd/vscode-clangd/assets/4633429/969bab6a-8086-410c-94c2-7c26acc621ac)

If this is considered desired behavior I would appreciate if someone could tell me a workaround to shut it off locally (preferably while maintaining the strikethrough behavior for deprecated code) 

**System information**
Clangd version (from the log, or `clangd --version`): 16.0.2
clangd extension version: v0.1.24

**Comment by HighCommander4** [comment]
So one way to disable it in vscode is to set `"editor.showDeprecated"` to false.

> preferably while maintaining the strikethrough behavior for deprecated code

Does the "deprecated code" situation apply to C++? If yes, could you show an example?

If not, then scoping the setting to just C++ should do it:

```json
"[cpp]": {
  "editor.showDeprecated": false
}
```

**Comment by arijun** [comment]
> Does the "deprecated code" situation apply to C++? If yes, could you show an example?

Yes, some of our teams will use deprecated to indicate there is a newer or more comprehensive function to use, without breaking existing usages. Here is a toy example:
![image](https://github.com/clangd/vscode-clangd/assets/4633429/e6999b66-cbb5-4a8a-9706-842223c5f73b)
The oldApiCall is correctly struck-through here.

In the interim I am disabling those lints so it's not a dealbreaker, but it would be nice to have them for code I have control over without making code not under my control unreadable.

**Comment by HighCommander4** [comment]
Ok, so that one is a `-Wdeprecated` diagnostic.

So, is your suggestion that clangd should send a `Deprecated` diagnostic tag for the `-Wdeprecated` diagnostic, but not for the clang-tidy `modernize-` diagnostics?

Conceptually the two seem similar ("you're using an old thing, you could be using a newer thing instead").

**Comment by arijun** [comment]
You're not wrong, although explicitly marking something as deprecated does seem like a stronger signal than the `modernize-` lints. If I call something that relies on deprecated code, I often have to make sure the deprecated calls don't have flaws that will affect me, something I never have to worry about with `modernize-`.

Truthfully, though, the main issue with the strikethrough comes from my example of typedef'ed enums, where it obscures important information. And that could probably be solved by changing the highlight region to just be on the `typedef` itself, or something similar. That's under the purview of clangd (I assume) so I will open a ticket there.

Until that gets changed there (if it does), would it be worth removing the strikethrough for `modernize-use-using`? I imagine I wouldn't be the only user with this issue.

**Comment by HighCommander4** [comment]
> Truthfully, though, the main issue with the strikethrough comes from my example of typedef'ed enums, where it obscures important information. And that could probably be solved by changing the highlight region to just be on the `typedef` itself, or something similar. 

Yeah, narrowing the diagnostic range to just the `typedef` keyword itself would make sense to me.

> That's under the purview of clangd (I assume) so I will open a ticket there.

The diagnostic is produced by clang-tidy, clangd just passes it through without modifying most properties like the diagnostic range, so the place to file that issue would be https://github.com/llvm/llvm-project/issues which is where clang-tidy issues are tracked. (Please feel free to drop a link to it here once you've filed it.)

> Until that gets changed there (if it does), would it be worth removing the strikethrough for `modernize-use-using`? I imagine I wouldn't be the only user with this issue.

cc @sammccall, who wrote the [patch](https://reviews.llvm.org/D141537) to use the `Deprecated` diagnostic tag for `modernize-` diagnostics, for thoughts on this.

I don't have a strong opinion either way. If this were to come up in a project I'm working on, my approach would be to use a `.clang-tidy` metadata file to disable the `modernize-use-using` diagnostic in subdirectories where I'm not planning to promptly convert the typedefs to usings.

**Comment by arijun** [comment]
clang-tidy issue: https://github.com/llvm/llvm-project/issues/62702

**Comment by sam-mccall** [comment]
The `Deprecated` label still seems semantically appropriate to me, given that the user has explicitly opted into wanting to modernize such uses - these checks are all off by default.

The end-to-end UX is poor here but there are many loosely-coupled layers involved (clang diagnostics, clang-tidy, specific check, clangd, LSP, vscode UI) and I think it's reasonable to expect the occasional wart.

Practically I agree that reducing the diagnostic range for this particular check is a good fix: it's the use of `typedef` that's the problem, not the other parts of the declaration.

(If clangd's heuristics to pick a range are preferring the fixit range over the diag loc, maybe we can add SourceRange(Loc, Loc) to the diag to suppress)

**Comment by HighCommander4** [comment]
Not sure if this helps at all, but https://reviews.llvm.org/D154443 downgrades the severity of diagnostics with the `Deprecated` tag to [DiagnosticSeverity.Hint](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticSeverity).
